### ISSUE TITLE
Precondition - Check If the Loop has more than one non-effectively-final variables

### ISSUE THREAD

**Issue by mdarefin** [mdarefin/Convert-For-Each-Loop-to-Lambda-Expression-Eclipse-Plugin#10]
The body of the initial for loop does not have more than one reference to local non-effectively-final variables defined outside the loop. Loops having only one outside reference can be refactored if the side-effect can be externalized to a reduce operation.

**Comment by mdarefin** [comment]
need some details explanation

**Comment by khatchad** [comment]
Yeah this might be a bit tricky. Can lambda expressions reference fields of class? What about final fields? You must be an expert in Java lambda expressions.

**Comment by mdarefin** [comment]
I believe the reference variable works, but not the final. need to do some search

**Comment by khatchad** [comment]
Yes, please research this. Also, check the netbeans implementation.

**Comment by mdarefin** [comment]
Sure, it's great suggestion

**Comment by mdarefin** [comment]
/*
     \* Precondition 2
     \* The body of the lambda does not refer Non-Effectively-Final(NEF) references
     \* that are defined outsed the loop. 
     \* This is because from within a lambda you cannot have references to a NEF.
     */
    protected Boolean More ...containsNEFs() {
        return visitor.hasNonEffectivelyFinalVars;
    }

this is the visit netbeans use to check Non Effective Final Variable, does eclipse has somthing like it in visotor

**Comment by khatchad** [comment]
How is `hasNonEffectivelyFinalVars` populated?

**Comment by mdarefin** [comment]
I think they have visit for this.
https://gist.github.com/mdarefin/c231c55a4622074107df

**Comment by khatchad** [comment]
Sorry. I have no idea what is going on here. Is this the NetBeans code you are sending me? If so, please give me a link to that code as opposed to your personal gist.

**Comment by mdarefin** [comment]
here is the link for netbeans code, for non-effective final variable they use containsNEFs()
http://grepcode.com/file/bits.netbeans.org/maven2/org.netbeans.modules/org-netbeans-modules-java-hints/RELEASE80-BETA/org/netbeans/modules/java/hints/jdk/mapreduce/PreconditionsChecker.java#PreconditionsChecker.containsNEFs%28%29

**Comment by khatchad** [comment]
Let's concentrate on #8.

**Comment by mdarefin** [comment]
Sure, I want to finish the preconditions by tomorrow.

**Comment by mdarefin** [comment]
for this in final variable to Enum conversion we also check for non-effective-final variable, can you send me the link how did you check it in the that pulg-in, it would help me to implement in this one
